File: src/Views/AdminClient/UserManagementView.cs
using CineComplex.Classes.Base;
using CineComplex.Classes;
using CineComplex.Interfaces;
using CineComplex.Models;
using CineComplex.ViewModels.FormViewModels;
using CineComplex.Views.FormViews;
using ConsoleTables;
using Microsoft.EntityFrameworkCore.Metadata.Internal;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CineComplex.Views.AdminClient
{
    public class UserManagementView : AViewBase<UserManagementView>, IView
    {
        private int _choice = 0;
        public int Choice { get => _choice; set => _choice = value; }
        public List<string> MenuList { get; set; }
        public void LoadMenuList()
        {
            Instance.MenuList = new List<string>()
            {
                "1. Show Users",
                "2. Create User",
                "3. Update User",
                "4. Block User",
                "5. Delete User",
                "6. Show User Bookings",
                "7. Exit"
            };
        }

        public void View()
        {
            Instance.LoadMenuList();

            do
            {
                Console.Clear();
                Console.WriteLine("\t----- !!! Salam Hindusthan !!! -----");
                Console.WriteLine("================================================");

                Console.WriteLine("\nManage Users - CineComplex");
                Console.WriteLine("-------------------------------------------------");

                Console.WriteLine();

                Console.WriteLine("\nMenu : ");
                Console.WriteLine("---------------");

                foreach (string instr in UserManagementView.Instance.MenuList)
                {
                    Console.WriteLine(instr);
                }

                Console.Write("Enter Your Choice : ");

                int.TryParse(Console.ReadLine(), out _choice);
                switch (Choice)
                {
                    case 1:
                        Instance.DisplayAllUsers();
                        break;
                    case 2:
                        SignUpFormView.Instance.View();
                        break;
                    case 3:
                        break;
                    case 4:
                        break;

                    case 5:
                        break;

                    case 6:
                        break;

                    default:
                        Console.WriteLine("Please enter the valid Choice .....");
                        break;
                }


            } while (Choice != UserManagementView.Instance.MenuList.Count);
        }

        public void DisplayAllUsers()
        {
            //Properties marked as NotMapped(hence set null) are also being used as columns for ConsoleTables
            Console.Clear();
            List<User> existingUsers = SQLInteraction.Db.Users.ToList();

            if (existingUsers.Any())
            {
                ConsoleTable table;
                int pageSize = 10;
                int currentPage = 0;
                bool hasMorePages = true;
                do
                {
                    Console.Clear();
                    Console.WriteLine("\t----- !!! Salam Hindusthan !!! -----");
                    Console.WriteLine("================================================");

                    Console.WriteLine("\nManage Users - CineComplex");
                    Console.WriteLine("-------------------------------------------------");

                    Console.WriteLine();

                    Console.WriteLine("\nUsers : ");
                    Console.WriteLine("---------------");

                    table = new ConsoleTable(new List<string>() { "Id", "UserName", "Contact", "Email" }.ToArray());
                    var pagedUsers = existingUsers.Skip(currentPage * pageSize).Take(pageSize);
                    foreach (User u in pagedUsers)
                    {
                        table.AddRow($"{u.Id}", $"{u.Username,-20}", $"{u.Contact,-5}", $"{u.Email,-25}");
                    }

                    if(pagedUsers.Count()!=0 && pagedUsers.Count()<10)
                    {
                        for(int i=0; i< 10-pagedUsers.Count();++i)
                        {
                            table.AddRow("","","","");
                        }
                    }

                    hasMorePages = pagedUsers.Count() == 0 ? false : true;
                    if (hasMorePages)
                    {
                        table.Write(Format.MarkDown);
                        Console.WriteLine($"Count {table.Rows.Count}");

                        currentPage++;
                      
                        Console.WriteLine("\nMenu : ");
                        Console.WriteLine("---------------");
                        Console.WriteLine("1. Press Enter to view the next page");
                        Console.WriteLine("2. Enter Id For Record Selection");
                        Console.WriteLine("3. Type 3 to exit");
                        Console.Write("Your Choice: ");

                        var input = Console.ReadLine();
                        if (input?.ToLower() == "3") 
                        { 
                            break; 
                        }

                        //else
                        //{
                            //Code to handle selected record
                        //}
                        
                    }
                } while (hasMorePages);

            }
            else
            {
                Console.WriteLine("No users found.");
            }
            //Console.ReadKey();
        }
    }
}

File: src/Views/AdminClient/UserManagementView.py
from tabulate import tabulate
import psycopg2
from typing import List

# PostgreSQL connection configuration
DB_CONFIG = {
    "dbname": "your_database",
    "user": "your_user",
    "password": "your_password",
    "host": "localhost",
    "port": "5432",
}

class User:
    def __init__(self, user_id, username, contact, email):
        self.user_id = user_id
        self.username = username
        self.contact = contact
        self.email = email

class UserManagementView:
    def __init__(self):
        self.choice = 0
        self.menu_list = []

    def get_connection(self):
        """Establish and return a PostgreSQL database connection."""
        return psycopg2.connect(**DB_CONFIG)

    def load_menu_list(self):
        """Load the menu options for user management."""
        self.menu_list = [
            "1. Show Users",
            "2. Create User",
            "3. Update User",
            "4. Block User",
            "5. Delete User",
            "6. Show User Bookings",
            "7. Exit"
        ]

    def view(self):
        """Display the User Management menu and handle user input."""
        self.load_menu_list()

        while True:
            print("\t----- !!! Salam Hindusthan !!! -----")
            print("================================================")
            print("\nManage Users - CineComplex")
            print("-------------------------------------------------")

            print("\nMenu:")
            print("---------------")

            for instr in self.menu_list:
                print(instr)

            choice = input("Enter Your Choice: ")
            try:
                self.choice = int(choice)
            except ValueError:
                print("Invalid choice, please try again.")
                continue

            if self.choice == 1:
                self.display_all_users()

            elif self.choice == 2:
                print("Creating user...")
                # Implement user creation logic

            elif self.choice == 3:
                print("Updating user...")
                # Implement user update logic

            elif self.choice == 4:
                print("Blocking user...")
                # Implement block user logic

            elif self.choice == 5:
                print("Deleting user...")
                # Implement delete user logic

            elif self.choice == 6:
                print("Displaying user bookings...")
                # Implement user bookings display logic

            elif self.choice == 7:
                print("Exiting User Management...")
                break

            else:
                print("Please enter a valid choice...")

    def display_all_users(self):
        """Display all users from the database in a paginated format."""
        users = self.get_users_from_db()

        if users:
            page_size = 10
            current_page = 0
            has_more_pages = True
            while has_more_pages:
                print("\nPage", current_page + 1)
                table_data = []
                paged_users = users[current_page * page_size:(current_page + 1) * page_size]

                for user in paged_users:
                    table_data.append([user.user_id, user.username, user.contact, user.email])

                print(tabulate(table_data, headers=["Id", "UserName", "Contact", "Email"], tablefmt="grid"))
                print(f"\nShowing {len(table_data)} users out of {len(users)}")

                if len(paged_users) < page_size:
                    has_more_pages = False

                if has_more_pages:
                    print("\n1. Press Enter to view the next page")
                    print("2. Enter Id for Record Selection")
                    print("3. Type 3 to exit")
                    choice = input("Your choice: ")

                    if choice == "3":
                        break
                    elif choice == "2":
                        record_id = input("Enter the User ID to select the record: ")
                        # Logic to handle the selected record
                    else:
                        current_page += 1
                else:
                    break
        else:
            print("No users found.")

    def get_users_from_db(self) -> List[User]:
        """Fetch users dynamically from PostgreSQL database."""
        users = []
        try:
            conn = self.get_connection()
            cursor = conn.cursor()
            cursor.execute("SELECT id, username, contact, email FROM users;")
            rows = cursor.fetchall()
            for row in rows:
                users.append(User(*row))
            cursor.close()
            conn.close()
        except psycopg2.Error as e:
            print(f"Database error: {e}")
        return users

# Example usage:
if __name__ == "__main__":
    user_management_view = UserManagementView()
    user_management_view.view()
